tools/jnimarshalmethod-gen/App.cs

```diff
@@ -221,6 +221,22 @@ static TypeBuilder GetTypeBuilder (ModuleBuilder mb, Type type)
 			return tb;
 		}
 
+		class MethodsComparer : IComparer<MethodInfo>
+		{
+			public int Compare (MethodInfo a, MethodInfo b)
+			{
+				bool aContainsDot = a.Name.IndexOf ('.') != -1;
```

Comment:
This is not an adequate way to determine if a `MethodInfo` is for an explicit interface implementation, [as mentioned elsewhere](https://github.com/xamarin/xamarin-android/pull/861#issuecomment-329808782).

For example -- and you made me dig up F# just to do this?! ;-) -- consider this F# declaration which implements the `System.IServiceProvider` interface:

```fsharp
open System;

type MyServiceProvider() =
  interface IServiceProvider with
    member this.GetService(serviceType : Type): System.Object = null
```

Compile, then disassemble, and the method we see is:

```cil
    // method line 2
    .method private virtual hidebysig newslot 
           instance default object 'System-IServiceProvider-GetService' (class [mscorlib]System.Type serviceType)  cil managed 
    {
        // Method begins at RVA 0x205c
	.override class [mscorlib]System.IServiceProvider::GetService
	// Code size 2 (0x2)
	.maxstack 8
	IL_0000:  ldnull 
	IL_0001:  ret 
    } // end of method MyServiceProvider::System-IServiceProvider-GetService
```

(Look mah, no `.`s!  It instead uses *dashes*!)

Which means simply checking for `.` won't work with F# code.

Instead, we need to use [`Type.GetInterfaceMap(Type)`](https://docs.microsoft.com/en-us/dotnet/api/system.type.getinterfacemap?view=netframework-4.7.2#System_Type_GetInterfaceMap_System_Type_).

`csharp` example of use:

```csharp
var t = typeof(List<int>);
var c = t.GetProperty("Count");

// `cm` implements ICollection<int>.Count.  How do we actually tell?
var cm = c.GetGetMethod ();

System.Reflection.InterfaceMapping map = cm.DeclaringType.GetInterfaceMap(typeof(ICollection<int>));
map.TargetMethods.Contains(cm); // true!
```

Alas, there doesn't appear to be anything as easy as Cecil's `MethodDefinition.Overrides`, which you used last time.

Also, calling `Type.GetInterfaceMap()` separately for every method will be terrible, perf-wise, so some form of machine would be needed.

That aside, simply checking for `.` is insufficient.  It *will* break.

Reply:
> This is not an adequate way to determine if a MethodInfo is for an explicit interface implementation, as mentioned elsewhere.

I didn't like the string comparison either, but because I thought the method was not overriden in that class, I didn't find better solution and stopped thinking about it at that point. My bad. The class has the methods with override as I found later (it was partial class split into 2 source files).

Anyway. Thinking more about it, I guess the whole solution is actually wrong and we should have both marshal methods. Because we will be missing the marshal method, when someone calls the iface method from Java side.

I think we should just name the marshal methods differently and include the return type in the method name. So in this case we would have:
```
__<$>_jni_marshal_methods.n_append_Ljava_lang_CharSequence__Ljava_lang_Appendable
__<$>_jni_marshal_methods.n_append_Ljava_lang_CharSequence__Ljava_lang_CharBuffer
```
Note double '_' before the return type name. It would actually be enough to have it for the iface method I guess:
```
__<$>_jni_marshal_methods.n_append_Ljava_lang_CharSequence__Ljava_lang_Appendable
__<$>_jni_marshal_methods.n_append_Ljava_lang_CharSequence
```
Then I hope it should be reachable from Java side as the registration signature contains the return type. Does it look OK to you? If so then I should probably try it to confirm it works.

Reply:
Thinking even more about it, we might not need to change it here at all as the registration class doesn't go thru C# and on IL level it should be OK. We should probably just check signatures from the register attribute in the linker, when updating the registration class.

Reply:
I have tried to update linker to update the registration methods with duplicates. There were issues with cecil usage, so resolved back to not generate the "duplicate" methods in the generator.